src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -2394,12 +2386,56 @@ void CodeGen::genPOPCNTIntrinsic(GenTreeHWIntrinsic* node)
     assert(node->gtHWIntrinsicId == NI_POPCNT_PopCount);
 
     genConsumeOperands(node);
-    // popcnt has false dependency on the target register on Intel Sandy Bridge, Haswell, and Skylake processors,
-    // so insert a `XOR target, target` to break the dependency via XOR triggering register renaming.
-    regNumber targetReg = node->gtRegNum;
-    getEmitter()->emitIns_R_R(INS_xor, EA_4BYTE, targetReg, targetReg);
-    genHWIntrinsic_R_RM(node, INS_popcnt, emitTypeSize(node->TypeGet()));
+    genXCNTIntrinsic(node, INS_popcnt);
     genProduceReg(node);
 }
 
+//------------------------------------------------------------------------
+// genXCNTIntrinsic: Generates the code for a lzcnt/tzcnt/popcnt hardware intrinsic node, breaks false dependencies on the target register
+//
+// Arguments:
+//    node - The hardware intrinsic node
+//    ins  - The instruction being generated
+//
+void CodeGen::genXCNTIntrinsic(GenTreeHWIntrinsic* node, instruction ins)
+{
+    // LZCNT/TZCNT/POPCNT have a false dependency on the target register on Intel Sandy Bridge, Haswell, and Skylake
+    // (POPCNT only) processors, so insert a `XOR target, target` to break the dependency via XOR triggering register
+    // renaming, but only if it's not an actual dependency.
+
+    GenTree*  op1        = node->gtGetOp1();
+    regNumber sourceReg1 = op1->gtRegNum;
+    regNumber sourceReg2 = REG_NA;
+
+    if (op1->isContained())
+    {
+        if (op1->OperIsHWIntrinsic())
```

Comment:
What's this for? I don't think that a contained operand on these instructions can be anything else than an indir.

Reply:
I got this check from `genHWIntrinsic_R_RM`. But I should only check for `isContained() && isIndir()`?